### PR TITLE
Read dstOverrides directly from service profiles

### PR DESCRIPTION
**Pull request by adleong**
Allow `dstOverrides` to be specified directly on a service profile as an alternative to using the SMI traffic split API.  This traffic split object:

```
apiVersion: split.smi-spec.io/v1alpha1
kind: TrafficSplit
metadata:
  name: books-error-inject
spec:
  service: books
  backends:
  - service: books
    weight: 80m
  - service: error-injector
    weight: 20m
```

is equivalent to this service profile:

```
apiVersion: linkerd.io/v1alpha2
kind: ServiceProfile
metadata:
  name: books.default.svc.cluster.local
  namespace: default
spec:
  dstOverrides:
  - service: error-injector
    weight: 20m
  - service: books
    weight: 80m
```

If a service has both a traffic split and a service profile that defines dstOverrides, the dstOverrides in the service profile are used and the traffic split is ignored.

Signed-off-by: Alex Leong <alex@buoyant.io>

### CODE REVIEW

**Comment by adleong** [comment]
@alpeb I'll update the template example to include a dstOverride.  Did you have any other changes in mind?  The validation uses the struct as a guide for unmarshalling so I don't think the field name needs to be accounted for.

I'm open to friendlier names.  We can pick whatever we think works best for Linkerd.  No need to necessarily match the SMI terminology here.  `split` feels a bit vague to me.

**Comment by alpeb** [comment]
@adleong you're right, that property is already accounted for. Persisting the service profile example from this issue returned a validation error and I assumed that was the cause, but the problem is just that `spec.routes` is missing.

**Comment by alpeb** [comment]
So far this has tested well for me. I also tried edge cases like starting with just an SP without a TS, or having both an SP and a TS and deleting the SP for the TS to be picked up :+1:

**Comment by adleong** [comment]
@grampelberg what do you think about this change?  From an internal architecture point of view it feels very natural because SMI traffic splits get converted into destination profiles under the hood anyway.

But it does increase the API surface area and makes it so there are two different ways of creating a traffic split.  It also may complicate future work related to a `linkerd get trafficsplit` command since traffic splits configured directly on the service profile would not have a corresponding SMI traffic split resource.

**Comment by grampelberg** [comment]
Originally, I had assumed that we would have a simple controller which takes TrafficSplit and rewrites to ServiceProfiles. I'm not a big fan of having two different places to configure it. What happens when someone changes the ServiceProfile setting but leaves TrafficSplit the same?

**Comment by adleong** [comment]
Service profile take precedence over traffic splits.

**Comment by grampelberg** [comment]
I don't particularly like adding extra config knobs that we and users need to reason about. I don't have any concrete issues though, from the product perspective adding a SMI-free version doesn't really get us anything ... but it also doesn't really hurt either.

tldr, 🤷‍♂

**Comment by adleong** [comment]
Aw, I wanted you to tell me no so I could blame someone else for taking away my candy 😜 

But I share your misgivings.  In general I'm a little unclear on how tightly coupled Linkerd and SMI should be and how deeply integrated SMI primitive should be in Linkerd as opposed to "native" Linkerd primitives like the service profile.  The fact that, for example, retries and traffic splits get configured in two different CRDs feels weird.

I'm going to close this for now until our vision about this comes into better focus.  We should keep this solution in mind for the future, though.
